test/functional/vimscript/system_spec.lua

```diff
@@ -691,4 +692,95 @@ describe('shell :!', function()
     feed('<CR>')
     helpers.expect_exit(command, 'qall!')
   end)
+
+  it(':.! should pipe line buffer to &shell only replacing the current line', function()
+    if is_os('win') then
+      -- Output of pwsh is lengthy.
+      -- Do not check the full string which is environment dependent
+      -- but the presence of critical words.
+      helpers.set_shell_powershell(true)
+      insert([[echo 'Hi, coconut!';
+exec zsh;]])
+      feed('gg')
+      command(':.!')
+      local found = string.find(curbuf_contents(), 'Hi, coconut!')
+      neq(found, nil)
+      found = string.find(curbuf_contents(), 'exec zsh;')
+      neq(found, nil)
+    else
+      insert([[find . 'src' -maxdepth 0 -type d;
+ls]])
+      feed('gg');
+      command(':.!')
+      local res = curbuf_contents()
+      eq(res, [[.
+src
+ls]])
+    end
+  end)
+
+  it(':%! should pipe buffer to &shell replacing the entire buffer', function()
+    if is_os('win') then
+      -- Output of pwsh is lengthy.
+      -- Do not check the full string which is environment dependent
+      -- but the presence of critical words.
+      helpers.set_shell_powershell(true)
+      insert([[dir;
+echo 'Hi, vimers!';]])
+      command(':%!')
+      local found = string.find(curbuf_contents(), 'Hi, vimers!')
+      neq(found, nil)
+      found = string.find(curbuf_contents(), 'Directory')
+      neq(found, nil)
+      found = string.find(curbuf_contents(), 'neovim')
+      neq(found, nil)
+    else
+      insert([[find . 'src' -maxdepth 0 -type d;
+echo 'hello world';]])
+      command(':%!')
+      local res = curbuf_contents()
+      eq(res, [[.
+src
+hello world]])
+    end
+  end)
+
+  it('! should filter motion to &shell by default on startup', function()
+    if is_os('win') then
+      helpers.set_shell_powershell(true)
+      insert([[echo 'Hi, coconut!';
+exec zsh;]])
+      feed('gg!w<CR>')
+      local found = string.find(curbuf_contents(), 'Hi, coconut!')
+      neq(found, nil)
+      found = string.find(curbuf_contents(), 'exec zsh;')
+      neq(found, nil)
+    else
+      insert([[echo 'hello world']])
+      feed('!w<CR>')
+      local res = curbuf_contents()
+      eq(res, 'hello world')
+    end
+  end)
+
+  it('! should filter motion to &shell by default on startup but remember the change of {filter} program', function()
+    if not helpers.funcs.has('python3') then pending("missing `python3`") end
+    if not is_os('win') then
+      insert([[echo 'hello world';]])
+      feed('!w<CR>')
+      local res = curbuf_contents()
+      eq(res, 'hello world')
+      feed('ggdd')
+      insert([[print("coucou")
+if True:
+    print("Hello, Vimers")]])
+      feed('gg')
+      feed('!wpython3<CR>')
```

Comment:
to avoid depending on `python3`, it could use `nvim` as a filter?

```suggestion
      insert([[print('foo')]])
      feed('!w nvim -l -<CR>')  -- Executes "print('foo')" using Nvim Lua.
```

Reply:
That is a great idea! I did not think about this for a filter program, and I was annoyed to use python in this test.

Reply:
Should this be `w!` instead of `!w` ?

Reply:
What is the `w` in `!wpython3` ?

Reply:
Oh, it seems the `nvim` executable is not installed on the CI machines. Do you think it is a good idea to test with the compiled executable ? I will try this, to check the CI, but I am not sure if it is a good idea or not.